File: domain/src/main/java/com/ricamgar/notify/domain/reminder/usecase/AddOrUpdateReminder.java
package com.ricamgar.notify.domain.reminder.usecase;

import com.ricamgar.notify.domain.base.usecase.UseCase;
import com.ricamgar.notify.domain.reminder.model.Reminder;
import com.ricamgar.notify.domain.reminder.repository.RemindersRepository;

import javax.inject.Inject;
import javax.inject.Named;

import rx.Observable;
import rx.Scheduler;

public class AddOrUpdateReminder extends UseCase<Reminder> {

    private final RemindersRepository remindersRepository;
    private Reminder reminder;

    @Inject
    protected AddOrUpdateReminder(RemindersRepository remindersRepository,
                                  @Named(value = "ioThread") Scheduler executionThread,
                                  @Named(value = "mainThread") Scheduler postExecutionThread) {
        super(executionThread, postExecutionThread);
        this.remindersRepository = remindersRepository;
    }

    public AddOrUpdateReminder setReminder(Reminder reminder) {
        this.reminder = reminder;
        return this;
    }

    @Override
    protected Observable<Reminder> buildUseCaseObservable() {
        if (reminder == null) {
            return Observable.error(new IllegalArgumentException());
        }
        return remindersRepository.addOrUpdate(reminder);
    }
}

File: domain/src/main/java/com/ricamgar/notify/domain/reminder/usecase/AddOrUpdateReminder.kt
package com.ricamgar.notify.domain.reminder.usecase

import com.ricamgar.notify.domain.base.usecase.UseCase
import com.ricamgar.notify.domain.reminder.model.Reminder
import com.ricamgar.notify.domain.reminder.repository.RemindersRepository
import rx.Observable
import rx.Scheduler
import javax.inject.Inject
import javax.inject.Named

class AddOrUpdateReminder
@Inject
constructor(private val remindersRepository: RemindersRepository,
            @Named(value = "ioThread") executionThread: Scheduler,
            @Named(value = "mainThread") postExecutionThread: Scheduler)
                : UseCase<Reminder>(executionThread, postExecutionThread) {
    private var reminder: Reminder? = null

    fun setReminder(reminder: Reminder): AddOrUpdateReminder {
        this.reminder = reminder
        return this
    }

    override fun buildUseCaseObservable() = when (reminder) {
        null -> Observable.error(IllegalArgumentException())
        else -> remindersRepository.addOrUpdate(reminder)
    }
}
